webapp/lib/controller.dart

```diff
@@ -597,17 +634,72 @@ void command(UIAction action, Data data) {
         return;
       }
       if (!view.sendingManualMultiMessageUserConfirmation(oneoffReply.text, selectedConversations.length)) {
         log.verbose('User cancelled sending manual multi message reply: "${oneoffReply.text}"');
         return;
       }
-      sendMultiReply(oneoffReply, selectedConversations);
+      queueMultiReply(oneoffReply, selectedConversations);
       view.conversationPanelView.clearNewMessageBox();
       break;
+    case UIAction.sendQueuedMessage:
+      MessageData messageData = data;
+      model.Conversation conversation = conversations.firstWhere((c) => c.docId == messageData.conversationId);
+      var messageToSend = conversation.messages[messageData.messageIndex];
+      sendReply(messageToSend, conversation);
+      break;
+    case UIAction.cancelQueuedMessage:
+      MessageData messageData = data;
+      switch (viewState) {
+        case UIView.conversations:
+          var messageToCancel = activeConversation.messages[messageData.messageIndex];
+          if (messageToCancel == selectedMessage) {
+            command(UIAction.deselectMessage, messageData);
+          }
+          activeConversation.messages.removeAt(messageData.messageIndex);
+          view.conversationPanelView.removeMessageAtIndex(messageData.messageIndex);
+          if (outboxActiveConversation.docId == activeConversation.docId) {
+            view.outboxConversationPanelView.removeMessageAtIndex(messageData.messageIndex);
+          }
+          if (activeConversation.messages.where((m) => m.status == model.MessageStatus.inOutbox).length == 0) {
+            queuedMessagesPerConversation.removeWhere((c, _) => c.docId == activeConversation.docId);
+          }
+          outboxActiveConversation = updateOutboxViewForConversations(queuedMessagesPerConversation.keys.toSet(), updateList: true);
+          break;
+        case UIView.outbox:
+          outboxActiveConversation.messages.removeAt(messageData.messageIndex);
+          view.outboxConversationPanelView.removeMessageAtIndex(messageData.messageIndex);
+          if (outboxActiveConversation.docId == activeConversation.docId) {
+            view.conversationPanelView.removeMessageAtIndex(messageData.messageIndex);
+          }
+          if (outboxActiveConversation.messages.where((m) => m.status == model.MessageStatus.inOutbox).length == 0) {
+            queuedMessagesPerConversation.removeWhere((c, _) => c.docId == outboxActiveConversation.docId);
+          }
+          outboxActiveConversation = updateOutboxViewForConversations(queuedMessagesPerConversation.keys.toSet(), updateList: true);
+          break;
+      }
+      break;
+
+    case UIAction.sendAllQueuedMessages:
+      for (var conversation in queuedMessagesPerConversation.keys) {
+        for (var message in conversation.messages.where((m) => m.status == model.MessageStatus.inOutbox)) {
+          sendReply(message, conversation);
+        }
+      }
```

Comment:
`queuedMessagesPerConversation.clear();` before `break;` ?

Reply:
I was thinking it would be good to keep the messages in the outbox until they've been confirmed to have been sent successfully, and only then remove them. So I'd rather not clear the list here, but rely on `sendReply` to remove items from this list as it's receiving failure/confirmation notifications.

It's not so clear from the name of this map that it actually holds the conversations in the outbox, not just queued messages for sending (I changed the approach half way through). I'll rename it to make it a bit clearer what its role is, so thanks for highlighting it!

Reply:
@marianamarasoiu Ah! That make sense. Thanks for the explanation.